File: metavision/sdk/cv/python_samples/metavision_dense_optical_flow/metavision_dense_optical_flow.py
# Copyright (c) Prophesee S.A. - All Rights Reserved
#
# Subject to Prophesee Metavision Licensing Terms and Conditions ("License T&C's").
# You may not use this file except in compliance with these License T&C's.
# A copy of these License T&C's is located in the "licensing" folder accompanying this file.

"""
Code sample showing how to use Metavision SDK to display results of dense optical flow.
"""

import numpy as np
import os
import h5py
import math
from metavision_core.event_io import EventsIterator
from metavision_core.event_io import LiveReplayEventsIterator, is_live_camera
from metavision_sdk_core import OnDemandFrameGenerationAlgorithm
from metavision_sdk_cv import PlaneFittingFlowAlgorithm, TripletMatchingFlowAlgorithm, DenseFlowFrameGeneratorAlgorithm, SpatioTemporalContrastAlgorithm
from metavision_sdk_ui import EventLoop, BaseWindow, Window, UIAction, UIKeyEvent
from skvideo.io import FFmpegWriter
from enum import Enum


class FlowType(Enum):
    PlaneFitting = "PlaneFitting"
    TripletMatching = "TripletMatching"

    def __str__(self):
        return self.value


def parse_args():
    import argparse
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(description='Metavision Dense Optical Flow sample.',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    input_group = parser.add_argument_group(
        "Input", "Arguments related to input sequence.")
    input_group.add_argument('-i', '--input-raw-file', dest='input_path', default="",
                             help="Path to input RAW file. If not specified, the live stream of the first available camera is used. "
                             "If it's a camera serial number, it will try to open that camera instead.")
    input_group.add_argument('--replay_factor', type=float, default=1,
                             help="Replay Factor. If greater than 1.0 we replay with slow-motion, otherwise this is a speed-up over real-time.")
    input_group.add_argument('--dt-step', type=int, default=33333, dest="dt_step",
                             help="Time processing step (in us), used as iteration delta_t period, visualization framerate and accumulation time.")

    algo_settings_group = parser.add_argument_group(
        "Algo Settings", "Arguments related to algorithmic configuration.")
    algo_settings_group.add_argument("--flow-type", dest="flow_type", type=FlowType, choices=list(FlowType), default=FlowType.TripletMatching,
                                     help="Chosen type of dense flow algorithm to run")
    algo_settings_group.add_argument("-r", "--receptive-field-radius", dest="receptive_field_radius", type=float,  default=3,
                                     help="Radius of the receptive field used for flow estimation, in pixels.")
    algo_settings_group.add_argument("--min-flow", dest="min_flow_mag", type=float,  default=10.,
                                     help="Minimum observable flow magnitude, in px/s.")
    algo_settings_group.add_argument("--max-flow", dest="max_flow_mag", type=float,  default=1000.,
                                     help="Maximum observable flow magnitude, in px/s.")
    algo_settings_group.add_argument("--stc-filter-thr", dest="stc_filter_thr", type=int,  default=40000,
                                     help="Length of the time window for filtering (in us).")
    algo_settings_group.add_argument("--visu-scale", dest="visualization_flow_scale", type=float,  default=-1,
                                     help="Flow magnitude used to scale the upper bound of the flow visualization, in px/s. If negative, will use 20\% of maximum flow magnitude.")

    output_flow_group = parser.add_argument_group(
        "Output flow", "Arguments related to output optical flow.")
    output_flow_group.add_argument(
        "--output-sparse-npy-filename", dest="output_sparse_npy_filename",
        help="If provided, the predictions will be saved as numpy structured array of EventOpticalFlow. In this "
        "format, the flow vx and vy are expressed in pixels per second.")
    output_flow_group.add_argument(
        "--output-dense-h5-filename", dest="output_dense_h5_filename",
        help="If provided, the predictions will be saved as a sequence of dense flow in HDF5 data. The flows are "
        "averaged pixelwise over timeslices of --accumulation-time-us. The dense flow is expressed in terms of "
        "pixels per timeslice (of duration accumulation-time-us), not in pixels per second.")
    output_flow_group.add_argument(
        '-o', '--out-video', dest='out_video', type=str, default="",
        help="Path to an output AVI file to save the resulting video.")
    output_flow_group.add_argument(
        '--fps', dest='fps', type=int, default=25,
        help="replay fps of output video")

    args = parser.parse_args()

    if args.output_sparse_npy_filename:
        assert not os.path.exists(args.output_sparse_npy_filename)
    if args.output_dense_h5_filename:
        assert not os.path.exists(args.output_dense_h5_filename)
    if args.visualization_flow_scale <= 0:
        args.visualization_flow_scale = args.max_flow_mag / 5

    return args


def accumulate_estimated_flow_data(args, width, height, processing_ts, flow_buffer, all_flow_events, all_dense_flows_start_ts, all_dense_flows_end_ts, all_dense_flows):
    """Accumulates estimated flow data in buffers to dump estimation results"""
    if args.output_sparse_npy_filename:
        all_flow_events.append(flow_buffer.numpy().copy())
    if args.output_dense_h5_filename:
        all_dense_flows_start_ts.append(
            processing_ts - args.dt_step)
        all_dense_flows_end_ts.append(processing_ts)
        flow_np = flow_buffer.numpy()
        if flow_np.size == 0:
            all_dense_flows.append(
                np.zeros((2, height, width), dtype=np.float32))
        else:
            xs, ys, vx, vy = flow_np["x"], flow_np["y"], flow_np["vx"], flow_np["vy"]
            coords = np.stack((ys, xs))
            abs_coords = np.ravel_multi_index(coords, (height, width))
            counts = np.bincount(abs_coords, weights=np.ones(flow_np.size),
                                 minlength=height*width).reshape(height, width)
            flow_x = np.bincount(
                abs_coords, weights=vx, minlength=height*width).reshape(height, width)
            flow_y = np.bincount(
                abs_coords, weights=vy, minlength=height*width).reshape(height, width)
            mask_multiple_events = counts > 1
            flow_x[mask_multiple_events] /= counts[mask_multiple_events]
            flow_y[mask_multiple_events] /= counts[mask_multiple_events]

            # flow expressed in pixels per delta_t
            flow_x *= args.dt_step * 1e-6
            flow_y *= args.dt_step * 1e-6
            flow = np.stack((flow_x, flow_y)).astype(np.float32)
            all_dense_flows.append(flow)


def dump_estimated_flow_data(args, width, height, all_flow_events, all_dense_flows_start_ts, all_dense_flows_end_ts, all_dense_flows):
    """Write accumulated flow results to output files"""
    try:
        if args.output_sparse_npy_filename:
            print("Writing output file: ", args.output_sparse_npy_filename)
            all_flow_events = np.concatenate(all_flow_events)
            np.save(args.output_sparse_npy_filename, all_flow_events)
        if args.output_dense_h5_filename:
            print("Writing output file: ", args.output_dense_h5_filename)
            flow_start_ts = np.array(all_dense_flows_start_ts)
            flow_end_ts = np.array(all_dense_flows_end_ts)
            flows = np.stack(all_dense_flows)
            N = flow_start_ts.size
            assert flow_end_ts.size == N
            assert flows.shape == (N, 2, height, width)
            dirname = os.path.dirname(args.output_dense_h5_filename)
            if not os.path.isdir(dirname):
                os.makedirs(dirname)
            flow_h5 = h5py.File(args.output_dense_h5_filename, "w")
            flow_h5.create_dataset(
                "flow_start_ts", data=flow_start_ts, compression="gzip")
            flow_h5.create_dataset(
                "flow_end_ts", data=flow_end_ts, compression="gzip")
            flow_h5.create_dataset("flow", data=flows.astype(
                np.float32), compression="gzip")
            flow_h5["flow"].attrs["input_file_name"] = os.path.basename(
                args.input_path)
            flow_h5["flow"].attrs["checkpoint_path"] = "metavision_dense_optical_flow"
            flow_h5["flow"].attrs["event_input_height"] = height
            flow_h5["flow"].attrs["event_input_width"] = width
            flow_h5["flow"].attrs["delta_t"] = args.dt_step
            flow_h5.close()
    except Exception as e:
        print(e)
        raise


def main():
    """ Main """
    args = parse_args()

    # Events iterator on Camera or RAW file
    mv_iterator = EventsIterator(
        input_path=args.input_path, delta_t=args.dt_step)

    # Set ERC to 20Mev/s
    if hasattr(mv_iterator.reader, "device") and mv_iterator.reader.device and mv_iterator.reader.device.get_i_erc_module():
        erc_module = mv_iterator.reader.device.get_i_erc_module()
        erc_module.set_cd_event_rate(20000000)
        erc_module.enable(True)

    if args.replay_factor > 0 and not is_live_camera(args.input_path):
        mv_iterator = LiveReplayEventsIterator(
            mv_iterator, replay_factor=args.replay_factor)
    height, width = mv_iterator.get_size()  # Camera Geometry

    # Event Frame Generator
    event_frame_gen = OnDemandFrameGenerationAlgorithm(
        width, height, args.dt_step)

    # Dense Optical Flow Algorithm
    # The input receptive field radius represents the total area of the neighborhood that is used to estimate flow. We
    # use an algorithm-dependent heuristic to convert this into the search radius to be used for each algorithm.
    if args.flow_type == FlowType.PlaneFitting:
        radius = math.floor(args.receptive_field_radius)
        print(
            f"Instantiating PlaneFittingFlowAlgorithm with radius={radius}")
        flow_algo_planefitting = PlaneFittingFlowAlgorithm(
            width, height, radius, -1)
        flow_algo_tripletmatching = None
    else:
        radius = 0.5*args.receptive_field_radius
        print(
            f"Instantiating TripletMatchingFlowAlgorithm with radius={radius}, min_flow={args.min_flow_mag}, max_flow={args.max_flow_mag}")
        flow_algo_tripletmatching = TripletMatchingFlowAlgorithm(
            width, height, radius, args.min_flow_mag, args.max_flow_mag)
        flow_algo_planefitting = None
    flow_buffer = TripletMatchingFlowAlgorithm.get_empty_output_buffer()

    # Dense Flow Frame Generator
    flow_frame_gen = DenseFlowFrameGeneratorAlgorithm(
        width, height, args.visualization_flow_scale, DenseFlowFrameGeneratorAlgorithm.AccumulationPolicy.Last)
    flow_legend_img = np.zeros((100, 100, 3), np.uint8)
    flow_frame_gen.generate_legend_image(flow_legend_img)

    # STC filter
    print(
        f"Instantiating SpatioTemporalContrastAlgorithm with thresh={args.stc_filter_thr}")
    stc_filter = SpatioTemporalContrastAlgorithm(
        width, height, args.stc_filter_thr, True)
    events_buf = SpatioTemporalContrastAlgorithm.get_empty_output_buffer()

    all_flow_events = []
    all_dense_flows = []
    all_dense_flows_start_ts = []
    all_dense_flows_end_ts = []

    # Window - Graphical User Interface
    with Window(title="Metavision Dense Optical Flow", width=width, height=2*height, mode=BaseWindow.RenderMode.BGR) as window, Window(
            title="Flow Legend", width=flow_legend_img.shape[1], height=flow_legend_img.shape[0], mode=BaseWindow.RenderMode.BGR) as window_legend:
        if args.out_video:
            video_name = args.out_video + ".avi"
            writer = FFmpegWriter(video_name, inputdict={'-r': str(args.fps)}, outputdict={
                '-vcodec': 'libx264',
                '-r': str(args.fps)
            })

        def keyboard_cb(key, scancode, action, mods):
            if action != UIAction.RELEASE:
                return
            if key == UIKeyEvent.KEY_ESCAPE or key == UIKeyEvent.KEY_Q:
                window.set_close_flag()

        window.set_keyboard_callback(keyboard_cb)
        window_legend.set_keyboard_callback(keyboard_cb)

        cd_output_img = np.zeros((height, width, 3), np.uint8)
        flow_output_img = np.zeros((height, width, 3), np.uint8)
        processing_ts = mv_iterator.start_ts
        # Process events
        for evs in mv_iterator:
            processing_ts += mv_iterator.delta_t

            # Dispatch system events to the window
            EventLoop.poll_and_dispatch()

            # Filter Events using STC
            stc_filter.process_events(evs, events_buf)

            # Generate Frame of Events
            event_frame_gen.process_events(events_buf)
            event_frame_gen.generate(processing_ts, cd_output_img)

            # Estimate the flow events
            if args.flow_type == FlowType.PlaneFitting:
                flow_algo_planefitting.process_events(events_buf, flow_buffer)
            else:
                flow_algo_tripletmatching.process_events(
                    events_buf, flow_buffer)
            accumulate_estimated_flow_data(args, width, height, processing_ts, flow_buffer,
                                           all_flow_events, all_dense_flows_start_ts, all_dense_flows_end_ts, all_dense_flows)

            # Draw the flow events on top of the events
            flow_frame_gen.process_events(flow_buffer)
            flow_frame_gen.generate(flow_output_img)

            # Update the display
            output_img = np.vstack((cd_output_img, flow_output_img))
            window.show(output_img)
            window_legend.show(flow_legend_img)

            if args.out_video:
                writer.writeFrame(output_img.astype(np.uint8)[..., ::-1])

            if window.should_close() or window_legend.should_close():
                break

    if args.out_video:
        writer.close()
    dump_estimated_flow_data(args, width, height, all_flow_events,
                             all_dense_flows_start_ts, all_dense_flows_end_ts, all_dense_flows)


if __name__ == "__main__":
    main()

File: metavision/sdk/cv/samples/metavision_dense_optical_flow/metavision_dense_optical_flow.cpp
/**********************************************************************************************************************
 * Copyright (c) Prophesee S.A. - All Rights Reserved                                                                 *
 *                                                                                                                    *
 * Subject to Prophesee Metavision Licensing Terms and Conditions ("License T&C's").                                  *
 * You may not use this file except in compliance with these License T&C's.                                           *
 * A copy of these License T&C's is located in the "licensing" folder accompanying this file.                         *
 **********************************************************************************************************************/

// This code sample demonstrates how to create a pipeline displaying the results of dense optical flow algorithms.
#include <iostream>
#include <chrono>
#include <functional>
#include <unordered_map>
#include <boost/program_options.hpp>
#include <metavision/sdk/core/algorithms/event_buffer_reslicer_algorithm.h>
#include <metavision/sdk/core/algorithms/on_demand_frame_generation_algorithm.h>
#include <metavision/sdk/cv/algorithms/dense_flow_frame_generator_algorithm.h>
#include <metavision/sdk/cv/algorithms/spatio_temporal_contrast_algorithm.h>
#include <metavision/sdk/cv/algorithms/plane_fitting_flow_algorithm.h>
#include <metavision/sdk/cv/algorithms/triplet_matching_flow_algorithm.h>
#include <metavision/sdk/driver/camera.h>
#include <metavision/sdk/ui/utils/event_loop.h>
#include <metavision/sdk/ui/utils/window.h>

namespace po = boost::program_options;

enum class DenseFlowType { PlaneFitting, TripletMatching };

const std::unordered_map<DenseFlowType, std::string> kDenseFlowTypeToStr = {
    {DenseFlowType::PlaneFitting, "PlaneFitting"}, {DenseFlowType::TripletMatching, "TripletMatching"}};
const std::unordered_map<std::string, DenseFlowType> kStrToDenseFlowType = {
    {"PlaneFitting", DenseFlowType::PlaneFitting}, {"TripletMatching", DenseFlowType::TripletMatching}};

std::istream &operator>>(std::istream &is, DenseFlowType &type) {
    std::string s;
    is >> s;
    auto it = kStrToDenseFlowType.find(s);
    if (it == kStrToDenseFlowType.cend())
        throw std::runtime_error("Failed to convert string to DenseFlowType");
    type = it->second;
    return is;
}

std::ostream &operator<<(std::ostream &os, const DenseFlowType &type) {
    auto it = kDenseFlowTypeToStr.find(type);
    if (it == kDenseFlowTypeToStr.cend())
        throw std::runtime_error("Failed to convert DenseFlowType to string");
    os << it->second;
    return os;
}

struct UserArguments {
    std::string bias_file_path;
    uint32_t afk_min_freq   = 90;
    uint32_t afk_max_freq   = 120;
    bool afk_band_pass      = true;
    uint32_t stc_threshold  = 10000;
    uint32_t erc_event_rate = 20000000;
    std::string in_file_path;
    std::string out_avi_file_path;
    DenseFlowType flow_type;
    float receptive_field_radius;
    float min_flow_mag, max_flow_mag;
    bool benchmark  = false;
    bool no_display = false;
    Metavision::timestamp processing_period;
    float visualization_flow_scale;
};

void configure_live_camera(Metavision::Camera &camera, const po::variables_map &vm, const UserArguments &args,
                           bool &hardware_stc_enabled) {
    // Configure biases:
    if (!args.bias_file_path.empty()) {
        camera.biases().set_from_file(args.bias_file_path);
    }
    // Configure AFK
    if (vm.count("min-freq") || vm.count("max-freq")) {
        try {
            auto *afk = camera.get_device().get_facility<Metavision::I_AntiFlickerModule>();
            auto mode = args.afk_band_pass ? Metavision::I_AntiFlickerModule::AntiFlickerMode::BAND_PASS :
                                             Metavision::I_AntiFlickerModule::AntiFlickerMode::BAND_STOP;

            afk->set_filtering_mode(mode);
            afk->set_frequency_band(args.afk_min_freq, args.afk_max_freq);
            afk->enable(true);
        } catch (Metavision::CameraException &e) {
            MV_LOG_ERROR() << "AFK not supported on this camera (available on Gen4.1 sensors and newer).";
        }
    }
    // Configure STC
    if (vm.count("stc-threshold")) {
        try {
            auto *event_trail_filter = camera.get_device().get_facility<Metavision::I_EventTrailFilterModule>();
            event_trail_filter->set_type(Metavision::I_EventTrailFilterModule::Type::STC_CUT_TRAIL);
            event_trail_filter->set_threshold(args.stc_threshold);
            event_trail_filter->enable(true);
            hardware_stc_enabled = true;
        } catch (Metavision::CameraException &e) {
            MV_LOG_ERROR() << "STC not supported on this camera (available on Gen4.1 sensors and newer).";
        }
    }
    // Configure ERC
    try {
        auto *erc = camera.get_device().get_facility<Metavision::I_ErcModule>();
        erc->set_cd_event_rate(args.erc_event_rate);
        erc->enable(true);
    } catch (Metavision::CameraException &e) {
        MV_LOG_ERROR() << "ERC not supported on this camera (available on Gen4 sensors and newer).";
    }
}

int main(int argc, char *argv[]) {
    bool hardware_stc_enabled = false;
    UserArguments args;

    const std::string program_desc(
        "Code sample showing how to use Metavision SDK to display results of dense optical flow algorithms.\n");

    po::options_description options_desc("Options");
    // clang-format off
    options_desc.add_options()
        ("help,h", "Produce help message.")
        ("bias-file,b", po::value<std::string>(&args.bias_file_path), "Apply bias settings on the camera")
        ("min-freq,m", po::value<uint32_t>(&args.afk_min_freq), "AFK: Lowest frequency of the band (in Hz), for Gen4.1 sensors and newer")
        ("max-freq,M", po::value<uint32_t>(&args.afk_max_freq), "AFK: Highest frequency of the band (in Hz), for Gen4.1 sensors and newer")
        ("band-pass,s", po::value<bool>(&args.afk_band_pass), "AFK: True for band-pass (by default), and False for band-cut, for Gen4.1 sensors and newer")
        ("stc-threshold,t", po::value<uint32_t>(&args.stc_threshold), "STC: filtering threshold delay (in us), for Gen4.1 sensors and newer")
        ("erc-event-rate,e", po::value<uint32_t>(&args.erc_event_rate)->default_value(20000000), "ERC: ERC target event rate (in event/s), for Gen4 sensors and newer")
        ("input-file,i", po::value<std::string>(&args.in_file_path)->default_value(""), "Path to input file. If empty, will try to open the first available live camera.")
        ("output-avi-file,o", po::value<std::string>(&args.out_avi_file_path)->default_value(""), "Path to output AVI file.")
        ("flow-type", po::value<DenseFlowType>(&args.flow_type)->default_value(DenseFlowType::TripletMatching), "Chosen type of dense flow algorithm to run, in {PlaneFitting, TripletMatching}")
        ("receptive-field-radius,r", po::value<float>(&args.receptive_field_radius)->default_value(3), "Radius of the receptive field, in pixels, used for flow estimation and converted for each method into the relevant radius.")
        ("min-flow", po::value<float>(&args.min_flow_mag)->default_value(10), "Minimum observable flow magnitude, in px/s.")
        ("max-flow", po::value<float>(&args.max_flow_mag)->default_value(1000), "Maximum observable flow magnitude, in px/s.")
        ("benchmark", po::bool_switch(&args.benchmark), "Configure pipeline to skip costly visualizations to enable timing of the dense flow algorithm specifically")
        ("no-display,d", po::bool_switch(&args.no_display), "Disable output display window")
        ("processing-period", po::value<Metavision::timestamp>(&args.processing_period)->default_value(33333), "Period for slicing and processing events and for generating flow visualization, in us.")
        ("visu-scale", po::value<float>(&args.visualization_flow_scale)->default_value(-1), "Flow magnitude used to scale the upper bound of the flow visualization, in px/s. If negative, will use 20\% of maximum flow magnitude.")
        ;
    // clang-format on

    po::variables_map vm;
    try {
        po::store(po::command_line_parser(argc, argv).options(options_desc).run(), vm);
        po::notify(vm);
    } catch (po::error &e) {
        MV_LOG_ERROR() << program_desc;
        MV_LOG_ERROR() << options_desc;
        MV_LOG_ERROR() << "Parsing error:" << e.what();
        return 1;
    }

    if (vm.count("help")) {
        MV_LOG_INFO() << program_desc;
        MV_LOG_INFO() << options_desc;
        return 0;
    }

    if (args.visualization_flow_scale == -1)
        args.visualization_flow_scale = args.max_flow_mag / 5;

    const bool enable_display      = !args.no_display && !args.benchmark;
    const bool enable_video_writer = !args.out_avi_file_path.empty() && !args.benchmark;
    const bool enable_visu         = enable_display || enable_video_writer;

    // Initialize the camera
    Metavision::Camera camera;
    if (args.in_file_path.empty()) {
        try {
            camera = Metavision::Camera::from_first_available();
        } catch (const Metavision::CameraException &e) {
            MV_LOG_ERROR() << e.what();
            return 1;
        }
        configure_live_camera(camera, vm, args, hardware_stc_enabled);
    } else {
        try {
            camera = Metavision::Camera::from_file(args.in_file_path,
                                                   Metavision::FileConfigHints().real_time_playback(!args.benchmark));
        } catch (Metavision::CameraException &e) {
            MV_LOG_ERROR() << e.what();
            return 1;
        }
    }
    const unsigned short width  = camera.geometry().width();
    const unsigned short height = camera.geometry().height();

    /// Data flow:
    //
    //  0 (Cam) -->-- 1 (buffer reslicer) -->-- 2 (STC) ----------->-----------  3 (Flow)
    //                                          |                                |
    //                                          v                                v
    //                                          |                                |
    //                                          4 (CD Frame Generator)           5 (Flow Frame Generator)
    //                                          |                                |
    //                                          v                                v
    //                                          |------>------->-<--------<------|
    //                                                          |
    //                                                          v
    //                                                          |
    //                                          |------<-------<->-------->------|
    //                                          |                                |
    //                                          v                                v
    //                                          |                                |
    //                                          6 (Display)                      7 (Video writer)
    //

    std::unique_ptr<Metavision::SpatioTemporalContrastAlgorithm> stc_algo;
    if (!hardware_stc_enabled) {
        stc_algo = std::make_unique<Metavision::SpatioTemporalContrastAlgorithm>(width, height, 0);
        stc_algo->set_threshold(args.stc_threshold);
        MV_LOG_INFO() << "Instantiating software SpatioTemporalContrastAlgorithm with thresh=" << args.stc_threshold;
    }

    // Instantiate the flow estimation algorithms
    // The input receptive field radius represents the total area of the neighborhood that is used to estimate flow. We
    // use an algorithm-dependent heuristic to convert this into the search radius to be used for each algorithm.
    std::unique_ptr<Metavision::PlaneFittingFlowAlgorithm> plane_fitting_flow_algo;
    std::unique_ptr<Metavision::TripletMatchingFlowAlgorithm> triplet_matching_flow_algo;
    switch (args.flow_type) {
    case DenseFlowType::PlaneFitting: {
        const int radius = cvFloor(args.receptive_field_radius);
        MV_LOG_INFO() << "Instantiating PlaneFittingFlowAlgorithm with radius=" << radius;
        plane_fitting_flow_algo = std::make_unique<Metavision::PlaneFittingFlowAlgorithm>(width, height, radius, -1);
        break;
    }
    case DenseFlowType::TripletMatching: {
        const float radius = 0.5f * args.receptive_field_radius;
        MV_LOG_INFO() << "Instantiating TripletMatchingFlowAlgorithm with radius=" << radius;
        Metavision::TripletMatchingFlowAlgorithmConfig triplet_matching_config(radius, args.min_flow_mag,
                                                                               args.max_flow_mag);
        triplet_matching_flow_algo =
            std::make_unique<Metavision::TripletMatchingFlowAlgorithm>(width, height, triplet_matching_config);
        break;
    }
    default:
        throw std::runtime_error("Selected DenseFlowType is not implemented!");
    }

    //  Create a video from the generated frames
    std::unique_ptr<cv::VideoWriter> video_writer;
    if (enable_video_writer) {
        video_writer = std::make_unique<cv::VideoWriter>(
            args.out_avi_file_path, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'), 60, cv::Size(width, height));
    }

    // Initialize the window if not disabled
    std::unique_ptr<Metavision::Window> window_visu, window_legend;
    if (enable_display) {
        window_visu = std::make_unique<Metavision::Window>("Dense Optical Flow", width, height,
                                                           Metavision::BaseWindow::RenderMode::BGR);
        window_visu->set_keyboard_callback(
            [&window_visu](Metavision::UIKeyEvent key, int scancode, Metavision::UIAction action, int mods) {
                if (action == Metavision::UIAction::RELEASE &&
                    (key == Metavision::UIKeyEvent::KEY_ESCAPE || key == Metavision::UIKeyEvent::KEY_Q)) {
                    window_visu->set_close_flag();
                }
            });
        window_legend =
            std::make_unique<Metavision::Window>("Flow Legend", 100, 100, Metavision::BaseWindow::RenderMode::BGR);
        window_legend->set_keyboard_callback(
            [&window_legend](Metavision::UIKeyEvent key, int scancode, Metavision::UIAction action, int mods) {
                if (action == Metavision::UIAction::RELEASE &&
                    (key == Metavision::UIKeyEvent::KEY_ESCAPE || key == Metavision::UIKeyEvent::KEY_Q)) {
                    window_legend->set_close_flag();
                }
            });
    }

    std::unique_ptr<Metavision::OnDemandFrameGenerationAlgorithm> cd_framer;
    std::unique_ptr<Metavision::DenseFlowFrameGeneratorAlgorithm> flow_framer;
    cv::Mat legend_frame;
    if (enable_visu) {
        cd_framer   = std::make_unique<Metavision::OnDemandFrameGenerationAlgorithm>(width, height, 0,
                                                                                   Metavision::ColorPalette::Dark);
        flow_framer = std::make_unique<Metavision::DenseFlowFrameGeneratorAlgorithm>(
            width, height, args.visualization_flow_scale,
            Metavision::DenseFlowFrameGeneratorAlgorithm::AccumulationPolicy::Last);
        if (enable_display) {
            flow_framer->generate_legend_image(legend_frame);
        }
    }

    cv::Mat combined_frame(2 * height, width, CV_8UC3);
    cv::Mat cd_frame   = combined_frame(cv::Rect(0, 0, width, height));
    cv::Mat flow_frame = combined_frame(cv::Rect(0, height, width, height));
    Metavision::EventBufferReslicerAlgorithm reslicer(
        [&](Metavision::EventBufferReslicerAlgorithm::ConditionStatus s, Metavision::timestamp t, std::size_t n) {
            if (!enable_visu)
                return;
            cd_framer->generate(t, cd_frame, false);
            flow_framer->generate(flow_frame, false);
            if (enable_display) {
                window_visu->show(combined_frame);
                window_legend->show(legend_frame);
            }
            if (enable_video_writer) {
                video_writer->write(combined_frame);
            }
        });
    reslicer.set_slicing_condition(
        Metavision::EventBufferReslicerAlgorithm::Condition::make_n_us(args.processing_period));

    Metavision::timestamp ts_first = -1, ts_last = -1;
    std::vector<Metavision::EventCD> cd_buffer;
    std::vector<Metavision::EventOpticalFlow> flow_buffer;
    auto reslicer_process_events_cb = [&](const Metavision::EventCD *begin, const Metavision::EventCD *end) {
        if (begin != end) {
            if (ts_first == -1)
                ts_first = begin->t;
            ts_last = std::prev(end)->t;
        }
        // Filter the events
        if (hardware_stc_enabled) {
            // When sensor's STC is available, we don't use software STC
            cd_buffer.insert(cd_buffer.end(), begin, end);
        } else {
            stc_algo->process_events(begin, end, std::back_inserter(cd_buffer));
        }
        // Run the flow algorithm
        switch (args.flow_type) {
        case DenseFlowType::PlaneFitting:
            plane_fitting_flow_algo->process_events(cd_buffer.cbegin(), cd_buffer.cend(),
                                                    std::back_inserter(flow_buffer));
            break;
        case DenseFlowType::TripletMatching:
            triplet_matching_flow_algo->process_events(cd_buffer.cbegin(), cd_buffer.cend(),
                                                       std::back_inserter(flow_buffer));
            break;
        default:
            throw std::runtime_error("Selected DenseFlowType is not implemented!");
        }
        // Process the visualization
        if (enable_visu) {
            cd_framer->process_events(cd_buffer.cbegin(), cd_buffer.cend());
            flow_framer->process_events(flow_buffer.cbegin(), flow_buffer.cend());
        }
        cd_buffer.clear();
        flow_buffer.clear();
    };

    // Vector of CD events and vector of OpticalFlow events to store the output of the two algorithms
    camera.cd().add_callback([&](const Metavision::EventCD *begin, const Metavision::EventCD *end) {
        reslicer.process_events(begin, end, reslicer_process_events_cb);
    });

    const auto ts_system_start = std::chrono::high_resolution_clock::now();
    camera.start();
    while (camera.is_running()) {
        Metavision::EventLoop::poll_and_dispatch(20);
        if (enable_display && (window_visu->should_close() || window_legend->should_close())) {
            break;
        }
    }
    camera.stop();
    const auto ts_system_stop = std::chrono::high_resolution_clock::now();
    const auto elapsed        = std::chrono::duration_cast<std::chrono::milliseconds>(ts_system_stop - ts_system_start);

    MV_LOG_INFO() << "Ran in" << Metavision::Log::no_space << static_cast<float>(elapsed.count()) / 1000.f << "s";
    MV_LOG_INFO() << "Record duration " << Metavision::Log::no_space << static_cast<float>(ts_last - ts_first) / 1e6f
                  << "s";
    if (enable_video_writer) {
        video_writer->release();
        MV_LOG_INFO() << "Wrote video file to:" << args.out_avi_file_path;
    }

    return 0;
}
